Various features and improvements

saluzafa opened the pull request:
# Changelog
[feature] allow MacOS shortcuts on board (`Cmd+C`, `Cmd+V`, ...)
[feature] Entity assets picker
[feature] add config file `config.yaml` for configuring limits and database directory
[feature] allow ping & "go to" for movable objects and sync it with others players
[misc] add login.yaml to .gitignore

--------------------------------------------

# Features

## Configuration file

Server now uses a configuration file to define limits & database path constants.

## Entity Assets Picker

This allows GM to use special assets for their custom tokens (like players tokens).

![Screenshot 2023-09-03 at 18 43 51](https://github.com/doodlezucc/dungeonclub/assets/475791/8c77ae20-3dcb-4b3f-aab4-8f57b29430e1)
![Screenshot 2023-09-03 at 18 44 23](https://github.com/doodlezucc/dungeonclub/assets/475791/c7acfb4e-1c72-4ec8-b384-d477300c800d)

## Ping & Go To tools for movable objects

### Ping tool

When a GM does a ping, it makes the token "heartbeating" for 10 seconds

![Screenshot 2023-09-03 at 18 45 40](https://github.com/doodlezucc/dungeonclub/assets/475791/5bcd2332-4e67-4c98-919f-6568320599be)

### Go To tool

Allows a GM to focus all players' screen on a specific token.

![Screenshot 2023-09-03 at 18 45 55](https://github.com/doodlezucc/dungeonclub/assets/475791/63e6d2c0-743a-4ffa-9949-4d537cb85228)

pss-julien committed:
Various improvements

[feature] allow configuring limits from a YAML file
[feature] allow MacOS shortcuts on board
[feature] add support for entity assets picker
[feature] allow configuring database directory
[misc] add login.yaml to .gitignore

pss-julien committed:
allow ping & goto movable objects and sync it with others players

doodlezucc commented on bin/connections.dart:
```diff
@@ -386,6 +386,14 @@ class Connection extends Socket {
+      case a.GAME_MOVABLE_PING:
+        _requireInSession();
+        return notifyOthers(action, params);
+
```
Because these actions should only be called by the GM, I'd prefer `_requireOwnerOfSession()` as precondition.

doodlezucc commented on bin/data.dart:
```diff
@@ -27,7 +28,7 @@ import 'versioning.dart';
 class ServerData {
   static final _manualSaveWatch = Stopwatch();
-  static final directory = Directory('database');
+  static final directory = Directory(DungeonClubConfig.getDatabasePath() + "/database");
```
Use `path.join()` instead, so that a database path like `database_path: "example_db/"` with a trailing slash still works as expected.

doodlezucc commented on bin/server.dart:
```diff
@@ -311,7 +313,7 @@ class Server {
 
     final isDataFile =
-        path.startsWith('database/games') || path.startsWith('ambience/');
+        path.contains('database/games') || path.startsWith('ambience/');
```
I can't think of a good use case for this change, could you explain?

doodlezucc commented on bin/server.dart:
```diff
@@ -320,7 +322,7 @@ class Server {
 
     file ??= isDataFile
-        ? File(path)
+        ? File("${DungeonClubConfig.getDatabasePath()}/${path}")
```
Single quotes are generally used in this project, however the linter isn't enforcing recommended rules yet, so it's perfectly fine.

doodlezucc commented on config.yaml.example:
For variant files like this one (a variant to `config.yaml`), I'd rather retain the actual file extension, so that IDEs and file explorers automatically identify the content language. Maybe we could rename this to `example.config.yaml`?

doodlezucc commented on dev.sh:
This might be an oversight on my part. It would be better if the dev script always enabled mock accounts. Instead of including this shell script in the repo, I recommend editing `bin/dev.dart`. It should be possible to pass this mock account argument when launching the server isolate, but I'm not completely sure how.

doodlezucc commented on login.yaml.example:
Similar to the other example file, please use a name like `example.login.yaml` instead.

doodlezucc commented on lib/dungeon_club_config.dart:
The `lib/` directory is meant to include Dart code which is usable on both frontend and backend. Because you're using `dart:io` here to load the local config file, this would fit better inside the `bin/` directory. Also, I recommend not including the product name (Dungeon Club) in file names because that makes it sound kind of convoluted. Maybe we could move this file to `bin/config.dart`?

doodlezucc commented on lib/dungeon_club_config.dart:
```diff
@@ -0,0 +1,38 @@
+
+import 'package:yaml/yaml.dart';
+
+final _yamlConfig = loadYaml((new File("config.yaml")).readAsStringSync());
```
Dart doesn't require the `new` keyword to construct objects. Without it you could lose a pair of parantheses and make it more readable:

```dart
final _yamlConfig = loadYaml(File("config.yaml").readAsStringSync());
```

doodlezucc commented on scripts/setup-ubuntu-with-nginx.sh:
```diff
@@ -0,0 +1,37 @@
+sudo npm install -g sass
+
+# nginx setup
+sudo apt-get instll nginx python3-certbot-nginx
```
*install (small typo)

doodlezucc commented on rebuild.sh:
I'm going to have to be honest: Although these scripts are very nicely written, I don't think they serve a general purpose as part of the repository and should therefore not be included. (Re-)building the server executable is helpful in some cases, but releases are already automatically built via [GitHub Actions](https://github.com/doodlezucc/dungeonclub/blob/master/.github/workflows/cross-build.yaml).

doodlezucc commented on scripts/nginx.conf:
See my comment in `rebuild.sh` - I'd rather exclude this file from the repo as it's too niche of a use case.

doodlezucc commented on scripts/setup-ubuntu-with-nginx.sh:
See my comment in `rebuild.sh` - I'd rather exclude this file from the repo as it's too niche of a use case.

doodlezucc commented on tmp/.keep:
I assume this `tmp/` directory is used when setting up nginx? Temporary directories should not be included in version control.

doodlezucc commented on start.sh:
Starting the (re-)compiled executable doesn't have to be part of the development process/repository IMO.

doodlezucc commented on bin/autosave.dart:
```diff
@@ -2,12 +2,13 @@ import 'dart:io';
-final _backupDaily = Directory('database_backup/daily');
-final _backupWeeks = Directory('database_backup/weeks');
+final _backupDaily = Directory("${DungeonClubConfig.getDatabasePath()}/database_backup/daily");
+final _backupWeeks = Directory("${DungeonClubConfig.getDatabasePath()}/database_backup/weeks");
```
Prefer `path.join` for joining operation system paths, like so:

```dart
p.join(DungeonClubConfig.getDatabasePath(), 'database_backup/daily')
```

doodlezucc commented on .gitignore:
```diff
@@ -20,4 +21,10 @@ web/images/assets/*-preview.*
+config.yaml
+tmp/*
+!tmp/.keep
+login.yaml
```
I assume `db/` and `tmp/` are your dev database directories, please remove these entries from .gitignore as they are not being generated unless a specific custom database path is set.

doodlezucc commented on lib/limits.dart:
Because `limits.dart` is now dependent on a local configuration file, it's no longer used in frontend code. Therefore I'd move it to `bin/limits.dart` or simply move the contents of this file into your `lib/dungeon_club_config.dart` code.

doodlezucc commented on web/dart/session/board.dart:
```diff
@@ -452,13 +454,35 @@ class Board {
     _sendSelectedMovablesSnap();
   }
 
+  void _goToMovable() {
+    socket.sendAction(a.GAME_MOVABLE_GOTO, {
+      'movable': selected.first.id,
+    }).then((value) => selected.first.goTo());
+  }
+
+  void _pingMovable() {
+    socket.sendAction(a.GAME_MOVABLE_PING, {
+      'movable': selected.first.id,
+    }).then((value) => selected.first.ping());
+  }
+
```
Instead of `selected.first` I'd use `selected.active!`, which points to the "primary" selected token (visually displayed with a white ring around it).

doodlezucc replied:
Also, I personally like running the action on the sender's client side *before* sending it to the server, so that there's no visible delay in feedback.

```dart
void _goToMovable() {
  selected.active!.goTo();
  socket.sendAction(a.GAME_MOVABLE_GOTO, {
    'movable': selected.active!.id,
  });
}

void _pingMovable() {
  selected.active!.ping();
  socket.sendAction(a.GAME_MOVABLE_PING, {
    'movable': selected.active!.id,
  });
}
```

doodlezucc commented on web/dart/session/movable.dart:
```diff
@@ -316,6 +316,21 @@ class Movable extends InstanceComponent
+    });
+  }
+
+  void goTo() {
+    this.styleSelected = true;
+    board.animateTransformToToken(this);
+    Future.delayed(const Duration(seconds: 10), () {
+      this.styleSelected = false;
+    });
+  }
```
If a player had previously selected this specific token/movable, the styling would disappear after 10 seconds. Instead, it'd be more stable to add a different CSS class (e.g. "pinged") to the element for a short duration.

doodlezucc commented on web/dart/account.dart:
In a [case like this](https://dart.dev/effective-dart/design#avoid-starting-a-method-name-with-get), I recommend Dart's `get` syntax, which is [encouraged](https://dart.dev/effective-dart/design#do-use-getters-for-operations-that-conceptually-access-properties) for exposing object properties:

```dart
  int get mediaBytesPerCampaign {
    return limits["media_bytes_per_campaign"] as int;
  }
```

Or even shorter:

```dart
  int get mediaBytesPerCampaign => limits["media_bytes_per_campaign"] as int;
```

doodlezucc commented on web/dart/user.dart:
See my comment regarding the use of getters over methods on `web/dart/account.dart`.

Using getters, these methods should look something like this:

```dart
int get mediaBytesPerCampaign => account?.mediaBytesPerCampaign ?? 0;
int get prefabsPerCampaign => account?.prefabsPerCampaign ?? 0;
```

When editing this file, changes would have to be applied to dependent code, which could be done like so:

```dart
// currently: without getters
final mbLimit = user.getMediaBytesPerCampaign();

// preferred: using getters
final mbLimit = user.mediaBytesPerCampaign;
```

doodlezucc reviewed (changes requested):
Thank you so much for taking the time and submitting your changes to the repository! Seeing this PR in my inbox really brightened my day :)

I left some review comments on your commits which I hope are descriptive enough to provide feedback. There are some aspects I would like to change in the implementation, but overall I'm very thankful for your work.

Cheers!

doodlezucc commented:
Also, I'm not sure if they've changed their licensing, but as far as I know, only tokens made by [Ross McConnell](https://2minutetabletop.com/author/2mtt/) from 2-Minute Tabletop may be used in Dungeon Club. The ones you've added are made by David Wilson and unfortunately might have to be replaced.

When I first contacted 2-Minute Tabletop back in November 2021, that's what they told me:

> Just make sure you only use tokens made by Ross, and not tokens made by our other token artists, as those are not covered by the same licensing. The ones you have are good though, as that's all Ross's work!

> Tokens by HammertheShark and David are intended for personal use only. [...] The 'safe to use' tokens are here: https://tools.2minutetabletop.com/token-editor/index?author=rossmcconnell

pss-julien committed:
changes requested

see https://github.com/doodlezucc/dungeonclub/pull/18#pullrequestreview-1608595378

saluzafa commented:
Hello @doodlezucc !

Thanks a lot for all your time and reviews, very happy to get a feedback from you 😄 .
I'm sorry for all the code mistakes, it was my first time playing with Dart, it's a very nice language 😎 .

I took care of all the reviews you made, I hope I didn't miss something let me know if it's the case 😄 .
I also deleted all tokens made by David Wilson and downloaded all tokens from Ross McConnell ! 🎉 

Can you test the ping effect please? Not sure if the animation is good or not (I'm a very bad UX designer).

Cheers!

doodlezucc commented on web/dart/account.dart:
```diff
@@ -25,34 +25,13 @@ class Account {
+  int get mapsPerCampaign => limits["maps_per_campaign"] as int;
+  int get campaignsPerAccount => limits["campaigns_per_account"] as int;
+  int get playersPerCampaign => limits["players_per_campaign"] as int;
+  int get movablesPerScene => limits["movable_per_scene"] as int;
```
Should probably be `movables_per_scene`. Sorry, I didn't notice that in the previous commit.

doodlezucc replied:
Actually, because I'm trying to move away from the term "movable", you could instead write `tokens_per_scene` here and update the getter name accordingly.

saluzafa replied:
My bad :), it's fixed!

doodlezucc replied:
Oh, I misclicked - there's actually another comment on this line, but it's displayed as part of the review. Sorry about that!

pss-julien committed:
Update account.dart

doodlezucc reviewed (commented):
Wow, you're extremely fast! :D
I'll take a look at the styling of pinged tokens tomorrow. Apart from that, this PR looks good to me and is considered practically approved. Thanks so much for your contributions and for implementing the previous requested changes!

pss-julien committed:
use getter in DungeonClubConfig + rename movablesPerScene to tokensPerScene

saluzafa commented:
> Wow, you're extremely fast! :D

Thanks! :D

> I'll take a look at the styling of pinged tokens tomorrow. Apart from that, this PR looks good to me and is considered practically approved. Thanks so much for your contributions and for implementing the previous requested changes!

Your welcome :)

> Actually, because I'm trying to move away from the term "movable", you could instead write tokens_per_scene here and update the getter name accordingly.

- I renamed `movables_per_scene` to `tokens_per_scene` ;
- I also renamed the methods `movablesPerScene` to `tokensPerScene`.

Cheers!

doodlezucc commented:
Hi @saluzafa, the ping animation looks fine to me! Thanks for your work :)
